[RNMobile] Upgrade Gradle to `8.2.1` and AGP to `8.1.0`

fluiddot opened the pull request:
**Related PR:**
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/5989
* https://github.com/wordpress-mobile/WordPress-Android/pull/18775

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Apply needed changes to upgrade Gradle to version 8.2.1 and AGP (Android Gradle Plugin) to version ~~8.0.2~~ 8.1.0.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

This is needed to match the versions that will be used in the Android app. Otherwise, we won't be able to do composite builds, which are needed during the development workflow to test native changes without publishing the Android binaries.

Closes https://github.com/wordpress-mobile/gutenberg-mobile/issues/5990.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

This PR contains several changes introduced by @ParaskP7 in [the forked branch `build/upgrade-gradle-to-8.2.1-and-agp-to-8.0.2`](https://github.com/ParaskP7/gutenberg/tree/build/upgrade-gradle-to-8.2.1-and-agp-to-8.0.2). The main changes are:
- Upgrade Gradle and AGP in `react-native-aztec` package.
- Upgrade Gradle and AGP in `react-native-bridge` package.
- Upgrade Gradle and AGP in `react-native-editor` package.
- Extract versions into groups to make version management easier.
- Move package from Android manifest files to build files.
- Enable build config build feature.
- Include correct `react-native-aztec` project in bridge.
- Update Android CI jobs to use Java 17.
- Add newer version of `react-native-gradle-plugin` via package `@react-native/gradle-plugin`.
- Remove patch for `react-native-gradle` which is no longer need it.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
* Check that the Gutenberg demo app can be built.
* Smoke test the editor.

⚠️ **NOTE: AGP 8.1.0 requires Android Studio Giraffe | 2022.3.1 or above.**

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A

## Screenshots or screencast <!-- if applicable -->
N/A

ParaskP7 committed:
Build: Upgrade gradle to 8.1.1 for aztec (1/2)

Release Notes: https://docs.gradle.org/8.1.1/release-notes.html
Command: ./gradlew wrapper --gradle-version=8.1.1
--distribution-type=all

------------------------------------------------------------------------

FYI: You will notice that the associated 'gradle-wrapper.jar', 'gradlew'
and 'gradlew.bat' files are not updated. This is because of the breaking
changes that this upgrade introduces, including the 'AGP' upgrade and
its associated breaking changes (namespace, build option, etc). As such,
this Gradle upgrade command is only half done.

PS: When all the breaking changes are resolved, this Gradle command will
be issued again, which should fully complete this upgrade.

ParaskP7 committed:
Deps: Extract agp version to settings build gradle for aztec

ParaskP7 committed:
Deps: Extract publish to s3 version to settings build gradle for aztec

ParaskP7 committed:
Deps: Add versions groups to root build gradle for aztec

ParaskP7 committed:
Deps: Extract androidx legacy version to root build gradle for aztec

ParaskP7 committed:
Deps: Extract androidx gridlayout version to root build gradle for aztec

ParaskP7 committed:
Deps: Extract androidx cardview version to root build gradle for aztec

ParaskP7 committed:
Deps: Extract androidx appcompat version to root build gradle for aztec

ParaskP7 committed:
Deps: Extract androidx recyclerview version to root bld gradle for aztec

ParaskP7 committed:
Deps: Extract junit version to root build gradle for aztec

FYI: You will notice an additional 'aztec' related 'junit' version. This
is because the node modules are using an older version of JUnit, the
'4.12', that is, comparing to the newer '4.13' version that the 'aztec'
modules are using. As such, and in order to avoid behavior changes, two
such version were extracted to the root 'build.gradle' file with an
accompanying TODO for the future.

ParaskP7 committed:
Build: Upgrade agp to 8.0.2 for aztec

Release Notes: https://developer.android.com/build/releases/
gradle-plugin#agp-8-0-2

ParaskP7 committed:
Build: Move package from android manifest to build files for aztec

This is an AGP version '8.0' breaking change that requires 'namespace'
in module-level build scripts, see build failure below:

------------------------------------------------------------------------

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring root project
 '@wordpress_react-native-aztec'.
> Failed to notify project evaluation listener.
   > Could not create an instance of type
    com.android.build.api.variant.impl.LibraryVariantBuilderImpl.
      > Namespace not specified. Please specify a namespace in the
       module's build.gradle file like so:

        android {
            namespace 'com.example.namespace'
        }

        If the package attribute is specified in the source
         AndroidManifest.xml, it can be migrated automatically to the
         namespace value in the build.gradle file using the AGP Upgrade
         Assistant; please refer to https://developer.android.com/
         studio/build/agp-upgrade-assistant for more information.
   > Could not get unknown property 'bundleReleaseAar' for object of
    type org.gradle.api.publish.maven.internal.publication
    .DefaultMavenPublication.

------------------------------------------------------------------------

Explanation: "You must set the namespace in the module-level
'build.gradle.kts' file, rather than the manifest file."

For more info see: https://developer.android.com/build/releases/
gradle-plugin#

ParaskP7 committed:
Build: Move android related gradle properties to its own group for aztec

ParaskP7 committed:
Build: Upgrade gradle to 8.1.1 for aztec (2/2)

Release Notes: https://docs.gradle.org/8.1.1/release-notes.html
Command: ./gradlew wrapper --gradle-version=8.1.1
--distribution-type=all

------------------------------------------------------------------------

As per the previous d85ad7bda1a594de7d890d1d241561b3490e12df commit and
its description, this change is a follow-up upgrade, which fully
completes this Gradle upgrade.

ParaskP7 committed:
Build: Upgrade gradle to 8.2.1 for aztec

Release Notes: https://docs.gradle.org/8.2.1/release-notes.html
Command: ./gradlew wrapper --gradle-version=8.2.1
--distribution-type=all

ParaskP7 committed:
Build: Move android related gradle prop. to its own group for bridge

ParaskP7 committed:
Deps: Extract agp version to settings build gradle for bridge

ParaskP7 committed:
Deps: Extract publish to s3 version to settings build gradle for bridge

ParaskP7 committed:
Deps: Extract facebook fresco react native related version for bridge

ParaskP7 committed:
Deps: Extract google material version to root build gradle for bridge

ParaskP7 committed:
Deps: Extract wordpress utils version to root build gradle for bridge

ParaskP7 committed:
Deps: Extract junit version to root build gradle for bridge

ParaskP7 committed:
Build: Upgrade gradle to 8.1.1 for bridge (1/2)

Release Notes: https://docs.gradle.org/8.1.1/release-notes.html
Command: ./gradlew wrapper --gradle-version=8.1.1
--distribution-type=all

------------------------------------------------------------------------

FYI: You will notice that the associated 'gradle-wrapper.jar', 'gradlew'
and 'gradlew.bat' files are not updated. This is because of the breaking
changes that this upgrade introduces, including the 'AGP' upgrade and
its associated breaking changes (namespace, build option, etc). As such,
this Gradle upgrade command is only half done.

PS: When all the breaking changes are resolved, this Gradle command will
be issued again, which should fully complete this upgrade.

ParaskP7 committed:
Build: Upgrade agp to 8.0.2 for bridge

Release Notes: https://developer.android.com/build/releases/
gradle-plugin#agp-8-0-2

ParaskP7 committed:
Build: Migrate dsl properties from lint options to lint for aztec

FYI: This was suggested automatically using the AGP upgrade assistant.

ParaskP7 committed:
Build: Move package from android manifest to build files for bridge

This is an AGP version '8.0' breaking change that requires 'namespace'
in module-level build scripts, see build failure below:

------------------------------------------------------------------------

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring root project
 '@wordpress_react-native-bridge'.
> Failed to notify project evaluation listener.
   > Could not create an instance of type
    com.android.build.api.variant.impl.LibraryVariantBuilderImpl.
      > Namespace not specified. Please specify a namespace in the
       module's build.gradle file like so:

        android {
            namespace 'com.example.namespace'
        }

        If the package attribute is specified in the source
         AndroidManifest.xml, it can be migrated automatically to the
         namespace value in the build.gradle file using the AGP Upgrade
         Assistant; please refer to https://developer.android.com/
         studio/build/agp-upgrade-assistant for more information.
   > Could not get unknown property 'bundleReleaseAar' for object of
    type org.gradle.api.publish.maven.internal.publication
    .DefaultMavenPublication.

------------------------------------------------------------------------

Explanation: "You must set the namespace in the module-level
'build.gradle.kts' file, rather than the manifest file."

For more info see: https://developer.android.com/build/releases/
gradle-plugin#

ParaskP7 committed:
Build: Enable build config build feature for bridge

This is an AGP version '8.0' breaking change that changes the build
option default values, see build failure below:

------------------------------------------------------------------------

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':react-native-bridge'.
> Failed to notify project evaluation listener.
   > com.android.builder.errors.EvalIssueException: defaultConfig
    contains custom BuildConfig fields, but the feature is disabled.
   > Could not get unknown property 'bundleReleaseAar' for object of
    type org.gradle.api.publish.maven.internal.publication
    .DefaultMavenPublication.

------------------------------------------------------------------------

Explanation: "Starting with AGP 8.0, the default values for these flags
have changed to improve build performance. AGP 8.0 doesn't generate
'BuildConfig' by default. You need to specify this option using the DSL
in the projects where you need it."

For more info see: https://developer.android.com/build/releases/
gradle-plugin#default-changes

ParaskP7 committed:
Build: Upgrade gradle to 8.1.1 for bridge (2/2)

Release Notes: https://docs.gradle.org/8.1.1/release-notes.html
Command: ./gradlew wrapper --gradle-version=8.1.1
--distribution-type=all

------------------------------------------------------------------------

As per the previous 5ff4e6e6f754b33bd0f41634230fa27e4b8a8c26 commit and
its description, this change is a follow-up upgrade, which fully
completes this Gradle upgrade

ParaskP7 committed:
Build: Migrate dsl properties from lint options to lint for bridge

FYI: This was suggested automatically using the AGP upgrade assistant.

ParaskP7 committed:
Build: Upgrade gradle to 8.2.1 for bridge

Release Notes: https://docs.gradle.org/8.2.1/release-notes.html
Command: ./gradlew wrapper --gradle-version=8.2.1
--distribution-type=all

fluiddot committed:
Build: Upgrade gradle to 8.2.1 for editor

fluiddot committed:
Build: Upgrade agp to 8.0.2 for editor

fluiddot committed:
Build: Move package from android manifest to build files for editor

fluiddot committed:
Build: Enable build config build feature for editor

fluiddot committed:
Patch `react-native-gradle` package

fluiddot committed:
Include correct `react-native-aztec` project in bridge

fluiddot commented:
Based on the error in `React Native E2E Tests (Android)` job:

```
An exception occurred applying plugin request [id: 'com.android.library', version: '8.0.2']
> Failed to apply plugin 'com.android.internal.library'.
   > Android Gradle plugin requires Java 17 to run. You are currently using Java 11.
```

We also need to update the CI configuration to use Java 17.

ParaskP7 commented on packages/react-native-editor/android/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.4.2-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.2.1-all.zip
```
Warning (⚠️): I understand that you manually updated the `distributionUrl`. Actually, the correct way of updating Gradle is using not manually, but rather by using this Gradle wrapper command: `./gradlew wrapper --gradle-version=8.2.1 --distribution-type=all`

FYI: Try running this command a few times and until all, the `gradle-wrapper.jar`, `gradlew`
and `gradlew.bat` files get updated.

fluiddot replied:
Thanks for the call out 🙇 ! I originally changed it manually but ended up using the command `./gradlew wrapper --gradle-version=8.2.1 --distribution-type=all` after checking the Gradle documentation. Seems I forgot to push those changes when I created the PR, I'll do it now.

fluiddot replied:
This was fixed in https://github.com/WordPress/gutenberg/pull/52872/commits/1d6ffba9a370e9dd6e094a2f0b735567c6b1c357.

ParaskP7 replied:
Awesome, verified, thanks for that @fluiddot ! 🙇 ❤️ 🚀

ParaskP7 commented on packages/react-native-bridge/android/settings.gradle:
```diff
@@ -22,11 +24,14 @@ pluginManagement {
+}
+else {
+    include ':@wordpress_react-native-aztec'
+    project(':@wordpress_react-native-aztec').projectDir = new File(rootProject.projectDir, '../../react-native-aztec/android')
```
1. Praise (❤️): Thanks for making this change to include the correct `react-native-aztec` project in bridge, this LGTM. PS: Maybe @oguzkocer can take a look and sign this off as well.
2. Minor (🔍): Consider having this `} else {` in a one-liner format.

fluiddot replied:
This suggestion was applied in https://github.com/WordPress/gutenberg/pull/52872/commits/bf340a441117768db07d5df889b0eae8ddb49c52.

oguzkocer replied:
When we are publishing the `react-native-bridge` library, [we use the compiled artifact for `react-native-aztec`](https://github.com/WordPress/gutenberg/blob/rnmobile/0.69.4-upgrade-gradle-to-8.2.1-and-agp-to-8.0.2/packages/react-native-bridge/android/react-native-bridge/build.gradle#L122-L126), so we shouldn't need this conditional.

I think you if you remove the duplicate `react-native-aztec` project, it should work as expected. So, I suggest reverting the changes in this file and only remove the following:

```
include ':react-native-aztec'
project(':react-native-aztec').projectDir = new File(rootProject.projectDir, '../../../packages/react-native-aztec/android')
```

fluiddot replied:
> When we are publishing the `react-native-bridge` library, [we use the compiled artifact for `react-native-aztec`](https://github.com/WordPress/gutenberg/blob/rnmobile/0.69.4-upgrade-gradle-to-8.2.1-and-agp-to-8.0.2/packages/react-native-bridge/android/react-native-bridge/build.gradle#L122-L126), so we shouldn't need this conditional.

Good point. I'll remove including the Aztec project in the publishing case.

fluiddot replied:
This was fixed in https://github.com/WordPress/gutenberg/pull/52872/commits/66a02f94b91f429b77b2dc391bd41110dde00b38.

ParaskP7 replied:
Great, even better, thanks both @fluiddot and @oguzkocer ! 🙇 ❤️ 🚀

fluiddot replied:
> This was fixed in [66a02f9](https://github.com/WordPress/gutenberg/commit/66a02f94b91f429b77b2dc391bd41110dde00b38).

Following this comment, I'll mark this conversation as resolved.

ParaskP7 commented on patches/react-native-gradle-plugin+0.0.7.patch:
```diff
@@ -0,0 +1,38 @@
+ tasks.withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile> {
+     kotlinOptions {
+-        jvmTarget = JavaVersion.VERSION_11.majorVersion
++        jvmTarget = JavaVersion.VERSION_17.majorVersion
```
1. Question (❓): I am not sure why and how this patch got generated, where did you get this newer version of this file? 🤔 For example, I am seeing that the newest version of such [react-native-gradle-plugin](https://github.com/invertase/react-native-gradle-plugin/) is [this](https://github.com/invertase/react-native-gradle-plugin/blob/master/build.gradle).
2. If that was an arbitrary update, that is, just to get this working:
    1. Question (❓): Was updating `Kotlin` necessary, and if so, why `1.8.10`?
    2. Question (❓): Was updating `AGP` necessary, and if so, why `7.2.1`?
    3. Warning (⚠️): Why Java `sourceCompatibility`, `targetCompatibility` and Kotlin `jvmTarget` are all now pointing to Java 17 (aka `JavaVersion.VERSION_17`)? I am asking that because I do believe this will cause trouble on older versions of Android. Instead, and if at all possible, I recommend all of them to be targeting Java 8 (aka `JavaVersion.VERSION_1_8`).

fluiddot replied:
> 1. Question (❓): I am not sure why and how this patch got generated, where did you get this newer version of this file? 🤔

This patch was generated using the [`patch-package` NPM command](https://www.npmjs.com/package/patch-package). It basically generates a patch and applies it every time you install the dependencies. It's a good alternative when you need to tweak something from a package but you don't want to fork it.

> For example, I am seeing that the newest version of such [react-native-gradle-plugin](https://github.com/invertase/react-native-gradle-plugin/) is [this](https://github.com/invertase/react-native-gradle-plugin/blob/master/build.gradle).

I think that's a different plugin, the one we are using comes from the React Native repository:
https://github.com/facebook/react-native/tree/main/packages/react-native-gradle-plugin

In fact, I'm thinking that we might reference a newer version of this package to solve the issue 🤔.

> * Question (❓): Was updating `Kotlin` necessary, and if so, why `1.8.10`?

I updated it to solve the following error:
```
e: Incompatible classes were found in dependencies. Remove them from the classpath or use '-Xskip-metadata-version-check' to suppress errors
e: .gradle/caches/8.2.1/generated-gradle-jars/gradle-api-8.2.1.jar!/META-INF/configuration-cache.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: .gradle/caches/8.2.1/generated-gradle-jars/gradle-api-8.2.1.jar!/META-INF/platform-jvm.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: .gradle/wrapper/dists/gradle-8.2.1-all/d8pvvlun5bx6sdtwqhf8y9z4b/gradle-8.2.1/lib/kotlin-stdlib-1.8.20.jar!/META-INF/kotlin-stdlib-jdk7.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: .gradle/wrapper/dists/gradle-8.2.1-all/d8pvvlun5bx6sdtwqhf8y9z4b/gradle-8.2.1/lib/kotlin-stdlib-1.8.20.jar!/META-INF/kotlin-stdlib-jdk8.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: .gradle/wrapper/dists/gradle-8.2.1-all/d8pvvlun5bx6sdtwqhf8y9z4b/gradle-8.2.1/lib/kotlin-stdlib-1.8.20.jar!/META-INF/kotlin-stdlib.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: .gradle/wrapper/dists/gradle-8.2.1-all/d8pvvlun5bx6sdtwqhf8y9z4b/gradle-8.2.1/lib/kotlin-stdlib-common-1.8.20.jar!/META-INF/kotlin-stdlib-common.kotlin_module: Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
e: workspace/gutenberg-mobile/gutenberg/node_modules/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/AndroidConfiguration.kt: (19, 25): Class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.
The class is loaded from .gradle/wrapper/dists/gradle-8.2.1-all/d8pvvlun5bx6sdtwqhf8y9z4b/gradle-8.2.1/lib/kotlin-stdlib-1.8.20.jar!/kotlin/Unit.class
```

> * Question (❓): Was updating `AGP` necessary, and if so, why `7.2.1`?

I've tested again reverting this change and the build was successful. Probably I update it while testing and forgot to change it back, I'll revert it.

> * Warning (⚠️): Why Java `sourceCompatibility`, `targetCompatibility` and Kotlin `jvmTarget` are all now pointing to Java 17 (aka `JavaVersion.VERSION_17`)? I am asking that because I do believe this will cause trouble on older versions of Android. Instead, and if at all possible, I recommend all of them to be targeting Java 8 (aka `JavaVersion.VERSION_1_8`).

Ah, interesting. I changed it to solve the following error:
```
java.lang.IllegalArgumentException: Unknown Kotlin JVM target: 8
```

I noticed in the React Native repository that they changed this value to `VERSION_11` ([reference](https://github.com/cortinico/react-native/blob/bece6500f7d6374eab09917c9afd3bf6142f0b61/packages/react-native-gradle-plugin/build.gradle.kts#L53-L54)). Looks like the main reason was upgrading Kotlin to `1.8.0` and the fact that Kotlin changed the default version of the stdlib that ships with by default.

<hr>

I'm going to try to use a newer version of the `react-native-gradle` package, as seems the optimal approach.

fluiddot replied:
Regarding issues with `react-native-gradle-plugin`, my understanding is that using Gradle 8 requires also bump Kotlin to 1.8. Note that the version of the Gradle plugin that RN 0.69.4 uses points to an older version 1.6 ([reference](https://github.com/facebook/react-native/blob/4bdec975c95e8c46c804a8a00c73ebe696a6982b/packages/react-native-gradle-plugin/build.gradle.kts#L12)), this is likely producing the issue I shared on the first question.

What it's not totally clear to me is why we need to use Java 17, seems related to the fact that we use AGP 8 but I'm not totally sure 🤔

fluiddot replied:
I tested using the latest version of the React Native Gradle plugin (package `@react-native/gradle-plugin`) and the Gutenberg demo app is building successfully. I think for now we can take this approach (https://github.com/WordPress/gutenberg/pull/52872/commits/693c0fe639dc8bf116ab4398782c74f2bf29766e).

ParaskP7 replied:
👋 @fluiddot !

> This patch was generated using the [patch-package NPM command](https://www.npmjs.com/package/patch-package). It basically generates a patch and applies it every time you install the dependencies. It's a good alternative when you need to tweak something from a package but you don't want to fork it.

TIL about [patch-package NPM command](https://www.npmjs.com/package/patch-package) and thanks for the clarification on that! 🙏 

> I think that's a different plugin, the one we are using comes from the React Native repository:
https://github.com/facebook/react-native/tree/main/packages/react-native-gradle-plugin

Oh, another TIL about [facebook -> react-native-gradle-plugin](https://github.com/facebook/react-native/tree/main/packages/react-native-gradle-plugin) and many thanks for helping me see the RN light! 🙏 

> In fact, I'm thinking that we might reference a newer version of this package to solve the issue 🤔.

🤔 

> I updated it to solve the following error:

Thanks for the clarification on that! 🙏 

> I've tested again reverting this change and the build was successful. Probably I update it while testing and forgot to change it back, I'll revert it.

👍 but also, I was actually thinking if we could use `8.0.2`, probably not, right? 🤔 

> Ah, interesting. I changed it to solve the following error:

👍 

> I noticed in the React Native repository that they changed this value to VERSION_11 ([reference](https://github.com/cortinico/react-native/blob/bece6500f7d6374eab09917c9afd3bf6142f0b61/packages/react-native-gradle-plugin/build.gradle.kts#L53-L54)). Looks like the main reason was upgrading Kotlin to 1.8.0 and the fact that Kotlin changed the default version of the stdlib that ships with by default.

🤔 Yea, this [Java reference](https://github.com/cortinico/react-native/blob/bece6500f7d6374eab09917c9afd3bf6142f0b61/packages/react-native-gradle-plugin/build.gradle.kts#L53-L54) and this [Kotlin reference](https://github.com/cortinico/react-native/blob/bece6500f7d6374eab09917c9afd3bf6142f0b61/packages/react-native-gradle-plugin/build.gradle.kts#L59) are both pointing to `VERSION_11`, but still, I am not sure what that means for us. One the one hand we won't be using any `VERSION_11`, it seems that it is just that the plugin is using that, but on the other hand, everywhere else we are still using `VERSION_1_8`, thus, I bit confused on what's the correct action here, it seems using `VERSION_11` like that is okay, but I can't be sure, needs testing... 🤷 

> I'm going to try to use a newer version of the react-native-gradle package, as seems the optimal approach.

👍 

> Regarding issues with react-native-gradle-plugin, my understanding is that using Gradle 8 requires also bump Kotlin to 1.8.

As far as I am aware, Gradle 8 doesn't have such a requirement. For reference, check Aztec, which is on [Kotlin 1.6.10](https://github.com/wordpress-mobile/AztecEditor-Android/blob/trunk/settings.gradle#L2), and it can still be upgraded to Gradle 8 (see [relevant PR](https://github.com/wordpress-mobile/AztecEditor-Android/pull/1051)).

> Note that the version of the Gradle plugin that RN 0.69.4 uses points to an older version 1.6 ([reference](https://github.com/facebook/react-native/blob/4bdec975c95e8c46c804a8a00c73ebe696a6982b/packages/react-native-gradle-plugin/build.gradle.kts#L12)), this is likely producing the issue I shared on the first question.

👍 

> What it's not totally clear to me is why we need to use Java 17, seems related to the fact that we use AGP 8 but I'm not totally sure 🤔

Yeap, not sure as well, but also, I don't think this is related to AGP 8 (see `Aztec` example above).

> I tested using the latest version of the React Native Gradle plugin (package @react-native/gradle-plugin) and the Gutenberg demo app is building successfully. I think for now we can take this approach (https://github.com/WordPress/gutenberg/commit/693c0fe639dc8bf116ab4398782c74f2bf29766e).

Awesome and yes, this is indeed seems the safer choice here, no need for us to decide on Java, Kotlin and AGP version and compatibility explicitly, thus potentially making a mess out of it! 🎉 

-----

🙇 ❤️ 🚀 THANK YOU 🙇 ❤️ 🚀

fluiddot committed:
Use newer version of React Native Gradle plugin

fluiddot committed:
Make `else` statement one-line format in editor's `settings.gradle`

oguzkocer reviewed (commented):
@fluiddot The changes so far looks good to me. I've left [one comment](https://github.com/WordPress/gutenberg/pull/52872/files#r1272478413) about the `settings.gradle` changes for `react-native-bridge`.

For the GitHub action failure related to JDK 17, you can add the following step prior to the task execution so that CI has access to JDK 17:

```
- name: Set Up JDK 17
  uses: actions/setup-java@v3
  with:
    java-version: 17
    distribution: 'corretto'
```

You can see an example of this [in this PR](https://github.com/Automattic/android-dependency-catalog/pull/23).

---

We are also updating the [publish to S3 Gradle Plugin](https://github.com/Automattic/publish-to-s3-gradle-plugin) version to `0.8.0` for all our projects as part of this AGP update. Could you update [this](https://github.com/WordPress/gutenberg/pull/52872/files#diff-b3197640e56dbd29d30e72328f04d81757731f643eb856079e847c60a5e8494eR4) & [this](https://github.com/WordPress/gutenberg/pull/52872/files#diff-e0b35174a44c650c42f00c9b0ac7dc0a73fa7c19a1288f48c44301ff4e22065aR4) to reflect that? You'll also need to remove the line with `androidSourcesJar` from the publish block [here](https://github.com/WordPress/gutenberg/blob/rnmobile/0.69.4-upgrade-gradle-to-8.2.1-and-agp-to-8.0.2/packages/react-native-aztec/android/build.gradle#L130) & [here](https://github.com/WordPress/gutenberg/blob/rnmobile/0.69.4-upgrade-gradle-to-8.2.1-and-agp-to-8.0.2/packages/react-native-bridge/android/react-native-bridge/build.gradle#L137). Here is [an example commit](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/119/commits/864c02e4f7b59d7d83cd5bc7a6dc92fb41595c00) that encapsulates this change.

Hope this helps and let me know if you have any questions/concerns!

fluiddot committed:
Remove aztec project inclusion when publishing bridge binaries

fluiddot commented:
> @fluiddot The changes so far looks good to me. I've left [one comment](https://github.com/WordPress/gutenberg/pull/52872/files#r1272478413) about the `settings.gradle` changes for `react-native-bridge`.
> 
> For the GitHub action failure related to JDK 17, you can add the following step prior to the task execution so that CI has access to JDK 17:
> 
> ```
> - name: Set Up JDK 17
>   uses: actions/setup-java@v3
>   with:
>     java-version: 17
>     distribution: 'corretto'
> ```
> 
> You can see an example of this [in this PR](https://github.com/Automattic/android-dependency-catalog/pull/23).

Great, thanks @oguzkocer for sharing the update 🙇 !



> We are also updating the [publish to S3 Gradle Plugin](https://github.com/Automattic/publish-to-s3-gradle-plugin) version to `0.8.0` for all our projects as part of this AGP update. Could you update [this](https://github.com/WordPress/gutenberg/pull/52872/files#diff-b3197640e56dbd29d30e72328f04d81757731f643eb856079e847c60a5e8494eR4) & [this](https://github.com/WordPress/gutenberg/pull/52872/files#diff-e0b35174a44c650c42f00c9b0ac7dc0a73fa7c19a1288f48c44301ff4e22065aR4) to reflect that? You'll also need to remove the line with `androidSourcesJar` from the publish block [here](https://github.com/WordPress/gutenberg/blob/rnmobile/0.69.4-upgrade-gradle-to-8.2.1-and-agp-to-8.0.2/packages/react-native-aztec/android/build.gradle#L130) & [here](https://github.com/WordPress/gutenberg/blob/rnmobile/0.69.4-upgrade-gradle-to-8.2.1-and-agp-to-8.0.2/packages/react-native-bridge/android/react-native-bridge/build.gradle#L137). Here is [an example commit](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/119/commits/864c02e4f7b59d7d83cd5bc7a6dc92fb41595c00) that encapsulates this change.
> 
> Hope this helps and let me know if you have any questions/concerns!

Sure thing, I'll apply the updates you shared. Thanks!

fluiddot committed:
Use Java 17 in Android E2E tests

fluiddot committed:
Update publish-to-s3 Gradle plugin to 0.8.0

fluiddot committed:
Revert removing package from AndroidManifest

Revert removing package from AndroidManifest

fluiddot commented on packages/react-native-bridge/android/react-native-bridge/src/main/AndroidManifest.xml:
```diff
@@ -1,5 +1,5 @@
-<?xml version="1.0" encoding="utf-8"?>
-<manifest xmlns:android="http://schemas.android.com/apk/res/android">
+
+<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="org.wordpress.mobile.ReactNativeGutenbergBridge">
```
The React Native CLI fetches the `package` attribute from AndroidManifest files, so we can't remove it. Otherwise, we encounter the following error when installing the Pods for the demo project:

```
error Failed to build the app: No package name found. Found errors in gutenberg-mobile/gutenberg/node_modules/@wordpress/react-native-bridge/android/react-native-bridge/src/main/AndroidManifest.xml.
```

This can be tested by removing the `package` attribute and running the command `npx react-native config` (this command is used behind the scenes when installing the Pods).

ParaskP7 replied:
I think it is okay (safe) to keep this package definition here @fluiddot , that is, as long as we make sure to have the same namespace defined on the inner `react-native-bridge/build.gradle` level, just we already did. 👍 

PS: We just need to not forget to update one without the other, that is, if we ever need to do something like that anyway (but I doubt that this will be happening anytime soon). 🤷

fluiddot replied:
> I think it is okay (safe) to keep this package definition here @fluiddot , that is, as long as we make sure to have the same namespace defined on the inner `react-native-bridge/build.gradle` level, just we already did. 👍
> 
> PS: We just need to not forget to update one without the other, that is, if we ever need to do something like that anyway (but I doubt that this will be happening anytime soon). 🤷

Good point. I think having an inline comment in both `build.gradle` and `AndroidManifest.xml` files reminding this would be helpful to avoid forgetting that both need to be updated. I'll add this 💨 .

fluiddot replied:
I applied my above suggestion in https://github.com/WordPress/gutenberg/pull/52872/commits/67ca6bdc8251773153889ac9711f8adfba76a93b.

ParaskP7 reviewed (approved):
👋 @fluiddot !

From where I stand, everything LGTM, thank you so much for taking this to the finish line, you are a rockstar! 🙇 ❤️ 🚀 

As such, I am personally going to approve this, but I suggest having @oguzkocer taking a final look, getting his approval here too, that is, before we proceed with the JP/WPAndroid testing  part via this [JP/WPAndroid#18775](https://github.com/wordpress-mobile/WordPress-Android/pull/18775) client PR.

fluiddot commented:
> From where I stand, everything LGTM, thank you so much for taking this to the finish line, you are a rockstar! 🙇 ❤️ 🚀
> 
> As such, I am personally going to approve this, but I suggest having @oguzkocer taking a final look, getting his approval here too, that is, before we proceed with the JP/WPAndroid testing part via this [JP/WPAndroid#18775](https://github.com/wordpress-mobile/WordPress-Android/pull/18775) client PR.

Thanks, you're more than welcome @ParaskP7 😊. All CI jobs are passing, but I encountered issues related to CI jobs and Java 17 in the [Gutenberg Mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5989). I'm actively working on it, at least to solve the ones related to CircleCI. However, I'm a bit blocked with Buildkite so I'm wondering if you could take a look when you are available ([reference](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5989#discussion_r1273440605)). Thanks!

ParaskP7 commented:
> Thanks, you're more than welcome @ParaskP7 😊.

❤️ x ❤️ ^ ❤️ 

> All CI jobs are passing, but I encountered issues related to CI jobs and Java 17 in the https://github.com/wordpress-mobile/gutenberg-mobile/pull/5989. I'm actively working on it, at least to solve the ones related to CircleCI.

Yea, I am aware of that, just took a look... 😞 🤷 😞 

> However, I'm a bit blocked with Buildkite so I'm wondering if you could take a look when you are available (https://github.com/wordpress-mobile/gutenberg-mobile/pull/5989#discussion_r1273440605). Thanks!

As per my comment [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5989#discussion_r1273652857) I think it is best to have @oguzkocer take a quick look when he becomes available later today. In the meanwhile, unfortunately, I'll need to switch context and work on something different for the rest of my day today.

fluiddot committed:
Add inline comment to remind updating package name

This comment will avoid forgetting about updating the package name in all places needed.

fluiddot committed:
Add `.java-version` file to use Java 17

fluiddot commented on packages/react-native-aztec/android/build.gradle:
```diff
@@ -1,15 +1,28 @@
+        androidxRecyclerviewVersion = '1.1.0'
+
+        // test
+        junitAztecVersion = '4.13' // TODO: Align 'junit' versions.
```
@ParaskP7 should be this version aligned with [WordPress-Android](https://github.com/wordpress-mobile/WordPress-Android/blob/2e46ca59383c7eecba8060583a93765c4fb451c0/build.gradle#L91C5-L91C17) or [AztecEditor-Android](https://github.com/wordpress-mobile/AztecEditor-Android/blob/6a119a95d189a09fcaea27b4dd6b4413c1af69b4/build.gradle#L78)?

SiobhyB replied:
Just checking whether this question was answered in another thread, or is it still outstanding? If it's correct that we should use the WordPress-Android version, should we remove the `TODO`?

fluiddot replied:
Nope, it hasn't been answered yet. I assume aligning the versions is not critical and only a matter of keeping version consistency across libraries. In any case, I'll let @ParaskP7 and @oguzkocer share insights about this. In the meantime, I'll remove the `TODO`. If we need to follow up on this, we could do it in a GitHub issue.

fluiddot commented on packages/react-native-editor/package.json:
```diff
@@ -34,6 +34,7 @@
 		"@react-native-community/blur": "4.2.0",
 		"@react-native-community/slider": "https://raw.githubusercontent.com/wordpress-mobile/react-native-slider/v3.0.2-wp-4/react-native-community-slider-3.0.2-wp-4.tgz",
 		"@react-native-masked-view/masked-view": "0.2.6",
+		"@react-native/gradle-plugin": "0.72.11",
```
Originally, we tried patching the React Native Gradle plugin, as discussed [here](https://github.com/WordPress/gutenberg/pull/52872#discussion_r1272191197). But this turned out to be complex and hard to maintain. For this reason, the incompatibility with the Gradle plugin has been solved by adding a specific and newer version. This seems safe even though we use an older React Native version.

**NOTE:** The React Native Gradle Plugin is a transitive dependency of React Native.

fluiddot committed:
Remove `react-native-gradle` patch

fluiddot committed:
Merge branch 'trunk' into rnmobile/0.69.4-upgrade-gradle-to-8.2.1-and-agp-to-8.0.2

# Conflicts:
#	package-lock.json
#	packages/react-native-aztec/android/gradle/wrapper/gradle-wrapper.jar
#	packages/react-native-aztec/android/gradle/wrapper/gradle-wrapper.properties
#	packages/react-native-aztec/android/gradlew.bat
#	packages/react-native-bridge/android/gradle/wrapper/gradle-wrapper.jar
#	packages/react-native-bridge/android/gradle/wrapper/gradle-wrapper.properties
#	packages/react-native-bridge/android/gradlew.bat
#	packages/react-native-editor/android/gradle/wrapper/gradle-wrapper.jar
#	packages/react-native-editor/android/gradle/wrapper/gradle-wrapper.properties
#	packages/react-native-editor/android/gradlew.bat
#	packages/react-native-editor/android/settings.gradle
#	packages/react-native-editor/package.json

fluiddot committed:
Bump Android Gradle Plugin to `8.1.0`

fluiddot committed:
Update `react-native-editor` changelog

fluiddot commented:
Heads up that I've updated AGP to version to a newer version (`8.1.0`) in order to match WP-Android. Additionally, after updating the branch with `trunk`, I removed a patch we introduced for the RN upgrade `0.71.11` that is no longer necessary due to using a newer version of the React Native Gradle Plugin (https://github.com/WordPress/gutenberg/pull/52872/commits/665c586ce1ae71472e8ae27f6c11447ec1cbaabe).

fluiddot commented on patches/react-native-gradle-plugin+0.71.19.patch:
We no longer this patch as WP-Android will use an Android Gradle Plugin version above the one used by the React Native Gradle Plugin (`8.2.1`).

fluiddot commented on .github/workflows/enforce-pr-labels.yml:
Seems this file got automatically reformatted when updating with `trunk` and committing 🤔. I checked the format of other workflow files ([example](https://github.com/WordPress/gutenberg/blob/7fead25296de38e777218196657d6d899cc15b12/.github/workflows/performance.yml)) and now it's matching it, so looks like a safe change to apply.

cc @priethor as author of https://github.com/WordPress/gutenberg/pull/52760

priethor replied:
Aha, the linter didn't throw any errors, but it's true the spacing wasn't matching the other yaml files. Thanks for catching this, and feel free to push the changes! 👍

fluiddot reviewed (commented):
I've tested the following items and all succeeded:
* Publish `react-native-aztec` library locally ✅
* Publish `react-native-bridge` library locally ✅
* Build and run `react-native-editor` ✅
* Build and run WP-Android using [this PR](https://github.com/wordpress-mobile/WordPress-Android/pull/18775) and composite build ✅
**NOTE:** I had to locally bump AGP and update with `trunk` when using the WP-Android PR.

The PR is now ready to review and once we have a green light in https://github.com/wordpress-mobile/WordPress-Android/pull/18775, we could proceed to merge it. Thanks!

SiobhyB reviewed (approved):
@fluiddot, thank you for all the work on this! I confirmed that the demo app builds as expected for me and, looking through the changes, I only had one question that is likely not blocking. Based on that, as well as the questions/review from Petros and Oguz, going ahead to approve. 👏 👏 🎉

fluiddot committed:
Merge branch 'trunk' into rnmobile/0.69.4-upgrade-gradle-to-8.2.1-and-agp-to-8.0.2

# Conflicts:
#	.github/workflows/enforce-pr-labels.yml

fluiddot committed:
Remove TODO inline comment in `build.gradle`
